sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -219,17 +230,44 @@ Connection getConnection() throws Exception {
       return toBuilder().setRowMapper(rowMapper).build();
     }
 
+    public Read<T> withCoder(Coder<T> coder) {
+      checkNotNull(coder, "coder");
+      return toBuilder().setCoder(coder).build();
+    }
+
     @Override
     public PCollection<T> apply(PBegin input) {
       return input
           .apply(Create.of(getQuery()))
-          .apply(ParDo.of(new ReadFn<>(this)));
+          .apply(ParDo.of(new ReadFn<>(this))).setCoder(getCoder())
+          // generate a random key followed by a GroupByKey and then ungroup
+          // to prevent fusion
+          // see https://cloud.google.com/dataflow/service/dataflow-service-desc#preventing-fusion
+          // for details
+          .apply(ParDo.of(new DoFn<T, KV<Integer, T>>() {
+            private int randInt;
+            @Setup
+            public void setup() {
+              Random random = new Random();
+              randInt = random.nextInt();
```

Comment:
This uses the same key for the whole bundle, which is equivalent to not doing this at all. The random key should be generated in processElement so that different elements get distributed onto different keys.

Ideally (but not required): can you try running an example pipeline against, e.g., a local spark or flink cluster with a local database and confirm that it speeds up as you add more workers? (I wish we had some infrastructure for performance testing...) I'd suggest a scenario where the database contains a small number of records (e.g. 100), but processing of every record is, e.g., a 10-second sleep. It should take less than 1000 seconds to complete and should scale linearly with number of workers; but it'd be sufficient to test with, e.g., 10 workers and confirm that it takes about 100 seconds (modulo per-worker startup time).